### PR TITLE
Give ability to specify external MySQL instance

### LINKED ISSUE #109: External MySQL/Maria DB
Hi,
It's possible to externalize DB?
Use an external DB could be usefull

### PR DESCRIPTION
**Pull request by bebehei**
missing: docs
missing: tests with external DB
missing: docker-compose.yml changes

<del>Requires #110 to be merged and then needs to get rebased</del> (done)

Fixes #109 

Still much to do.

### CODE REVIEW

File: content/opt/setup/50-icinga2 [hunk 1]
```diff
@@ -28,15 +28,34 @@ fi
...
+if [ "${ICINGA2_IDO_MYSQL_HOST}" == "localhost" ]; then
+	mysql <<-END
+	  CREATE DATABASE IF NOT EXISTS ${ICINGA2_IDO_MYSQL_DATA};
+	  GRANT ALL
+	    ON ${ICINGA2_IDO_MYSQL_DATA}.*
+	    TO '${ICINGA2_IDO_MYSQL_USER}'@'${ICINGA2_IDO_MYSQL_HOST}'
+	    IDENTIFIED BY '${ICINGA2_IDO_MYSQL_PASS}';
+	END
+else
+	mysql \
```
Comment (bebehei) [inline content/opt/setup/50-icinga2]: This check for DB access is quite senseless, if we can't `exit 1` and `run-parts` will also exit with 1 if one of the scripts fail.
Also we might want to check for

```
SHOW GRANTS;
SHOW GRANTS FOR CURRENT_USER;
SHOW GRANTS FOR CURRENT_USER();
```

That we actually can seed the DB.

**Comment by bebehei** [comment]
@jjethwa This is heavily WIP here. I just want to track all my thoughts.

**Comment by gorootde** [comment]
What about these changes? No progress since august?

**Comment by bebehei** [comment]
> What about these changes? No progress since august?

Yes, there are no changes since the end of August. The changes got quite big and are much heavier to test. I came to the conclusion, that it might make more sense to make an external SQL server the default and drop the internal one. This requires many initial effort for this PR, but makes testing in future easier.

Anyway, I've dedicated my recent free time to other projects. Of this image, I currently run only one instance and therefore the personal need to fix this up is quite low.

**Comment by gorootde** [comment]
I was thinking about just changing the hostname in all configs from `localhost` to e.g. `mysql`. Following this approach, the only dynamic thing to be done on image build is adding an entry to `/etc/hosts` which will either point to `localhost` (default) or to the linked mysql container.
This should keep changes quite minimal. What do you think?

**Comment by bebehei** [comment]
It's not that easy. You have to require the docker image to bootstrap a full configuration even if it's started with empty configuration and data dirs. So on initial startup, the whole database has to get bootstrapped and also all MySQL users have to be created. For this, there is a superuser login required, which is available on local mysql server. With an external MySQL server you can't require a superuser login. It's not the point of having a shared DB instance and still giving every container full access.

So providing both MySQL inside and outside of the image, you have to catch different test cases. (e.g.: Wrong SQL user credentials vs uninitialized DB). While using MySQL inside the image, you can assume the credential stuff always works.

Let's proceed: To initialize our DBs, we just execute the schema files on every boot onto the DB. If it fails, ignore it. It'll only fail when the schema file has already been executed in this DB.

But if you would use now credentials, you can't simply ignore the error. You now have to handle the error and stop the whole container if the credentials are wrong, but ignore it if the schema is good.

---

All in All: In german, we say "The one who wants to achieve something, will find a path. The one who doesn't, will find reasons." If you read my comment, you'll identify me as the "reasons"-guy, but the problem boils down to a "throwing money (a.k.a. time) at it problem" and I haven't thrown enough time in.

**Comment by bebehei** [comment]
Good news: On my local machine, it works™️ with a separate `mariadb` container and a dedicated `MYSQL_ROOT_PASSWORD`.

Now the tricky part begins: Proving that it properly fails, when something goes wrong.

Still to do: docs

**Comment by bebehei** [comment]
@gorootde @jjethwa Will you volunteer and give me some Feedback, please?

I'm finished with the docs, the transition process from internal bundled DB to an additional container works well. If permissions are not granted for the user, it'll also fail. And the container is in good shape IMHO.

I'm pretty sure, this would suffice for most people, but I'm not convinced this will cover all usecases and common traps. So, a bit testing and feedback would be appreciated before merging it into master.

**Comment by jjethwa** [comment]
Sure thing, @bebehei 

I'll need a few days to run through some scenarios :)

**Comment by mludvig** [comment]
Hi does it work with pre-created usernames and databases? I should be using our central DB cluster (properly backed up, etc) and can get the the DBs and Users created but won't be given the root password. Is this possible with your changes?

**Comment by bebehei** [comment]
@mludvig Have you seen the attached additions to `README.md` in this PR?

**Comment by mludvig** [comment]
@bebehei I have now, thanks for the pointer and sorry for the noise :)

**Comment by bebehei** [comment]
@mludvig You don't have to excuse for this "noise". It's easy to overlook the docs, which get bundled in the PR. Albeit the changes are in the `README.md`, these aren't in the usual place to look for this file 😉 

I actually want especially your feedback, because your use case is exactly the case, why I've created this PR with so much details. Sorry, but you got elected by myself to be the test candidate for this PR. I especially want to get feedback, if the docs are understandable.

**Comment by mludvig** [comment]
Sure, give me a few days and I'll come back with a feedback. So far it looks good :)

**Comment by jjethwa** [comment]
Looking good for me too, but just give me another day or two to try a few weird things users might try 😛

**Comment by bebehei** [comment]
@jjethwa Take your time. I'd rather test it twice. In the last months, I've seen so much going south after merging a PR. I wish there would be some effective way to make automated integration testing possible.

**Comment by jjethwa** [comment]
@bebehei I think it should be possible using (a free) CircleCI or Travis CI account actually.  I have not had time to look into setting it up unfortunately 😞

**Comment by bebehei** [comment]
@jjethwa Well, that isn't the actual problem. Setting up a CI should be ez. I just have no clue, how to write a script for it, which  tests the stuff.

**Comment by jjethwa** [comment]
@bebehei Ah yeah, that's the hard part 😞

**Comment by jjethwa** [comment]
@bebehei I think I've managed to cover enough scenarios and everything seems to be working well enough.  If there are no other objections, let's merge this 😄

**Pull request merged**